### ISSUE TITLE
[Compaction]: TWCS major has 100% disk space amplification

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#16515]
HEAD: 65e42e4166cef5acafa7ce7f884d6cc2b55adc35

*Description*
We lived for a very long time assuming that this is what it is and there is nothing to do about it.
However I believe this can be improved.

**Comment by vladzcloudius** [comment]
I suggest to achieve that by separating the process of data segragation (splitting the data from a given sstable into pieces belonging to a new buckets configuration) and the compaction itself.

Let's recall reasons for running a major with TWCS data:
1) Schema configuration got changed, e.g. bucket size got changed.
2) (This should not happen but sometimes it does): Some data was updated or explicitly deleted and the user wants it to be gone/squashed ASAP.

So, both these need to be addressed.

I suggest to achieve this by doing the following while limiting the maximum disk space amplification:

------------------------
1) Go over all existing sstables and split them according to the new TWCS configuration. At each such step the disk space amplification is as the size of the corresponding sstable.

At this point all our data is segregated by the timestamp but there still might be duplications or tombstones with corresponding shadowed data.

2) Let's index sstable sets from each time bucket after (1) from 0 to N-1 starting from the one that has the oldest data.
    Let's denote sstables set from bucket i as SST(i)

```
for i = 0; i < N - 1; i++ {
   for j = i + 1; j < N; j++ {
      for every row in SST(i) if there is at least one overlapping row in SST(j) 
      compact SST(i) and SST(j) together and segregate the corresponding result. 
   }
}      
```
The disk space amplification of this step is as the max(sizeof(SST(i)) + sizeof(SST(j))) for every i and j between 0 and N-1. 
In simple words it's as large as the sum of two largest buckets.

At the end of step (2) there won't be any data duplication of shadowed data in any bucket and all data is going to be fully segregated and compacted.
3) (Optional): execute an STCS `major` on each SST(i).
------------------------

This algorithm clearly trades extra processing for disk space amplification because unlike the "classic" major compaction which has a linear complexity step (2) above has an O(N^2) complexity in a general case.

However with a TWCS there is a "twist": if the data and the use case are appropriate for TWCS and if the major is called close to the "violation event" of updating or deleting existing data then the duplicated data and/or explicit tombstones are going to be located in the bucket with the newest data (SST(N-1)).
Older buckets are expected to have no "intersections".

Hence for the majority of (SST(i), SST(j)) pairs from (2) the processing is going to be limited to reading and processing of the SST(j) index data which is supposed to be a relatively cheap operation.

And the only heavy compaction that would be left would be compactions of (SST(i),SST(N-1)) pairs for every i in (0;N-2) which in total has the same linear complexity as the major compaction today.

The bottom line:
* The proposed algorithm trades extra index reading for a very low disk space amplification (a typical buckets count is 10-20 buckets, which would result in 10-20% disk space amplification during the "new major")
* For a typical TWCS use case it is going to be slower but probably not by much because the amount of writes is going to be the same in the worst case and a lot less in cases when intersection is limited to a small number of buckets (e.g. if only data from a specific bucket was deleted/updated).

**Comment by raphaelsc** [comment]
I think it's lots of complexity for something that can be more efficiently addressed by using ICS in TWCS. The benefits of course won't be seen in OSS. 

Also I don't think we can perform this check (see below), today, without invoking index (nor we can do it efficiently, bloom filter operates on partitions keys, overlapping at token level is not deterministic for detecting row level overlapping; maybe we could use some cardinality estimator, e.g. hyperloglog, but I heard it can be expensive):
`for every row in SST(i) if there is at least one overlapping row in SST(j) `

Also I think the write amplification could be terrible if you have multiple sstables overlapping each other at their row range.

SSTable A overlaps with B, which overlaps with C and so on. 

To be really honest, I think the more appropriate solution is to introduce the concept of runs to TWCS, and that's done by applying ICS to it instead. Better cost benefit (yet more efficient).

So we fix this issue and https://github.com/scylladb/scylladb/issues/16514 too, killing two birds with one stone.

**Comment by vladzcloudius** [comment]
> I think it's lots of complexity for something that can be more efficiently addressed by using ICS in TWCS. The benefits of course won't be seen in OSS.
> 
> Also I don't think we can perform this check (see below), today, without invoking index (nor we can do it efficiently, bloom filter operates on partitions keys, overlapping at token level is not deterministic for detecting row level overlapping; maybe we could use some cardinality estimator, e.g. hyperloglog, but I heard it can be expensive): `for every row in SST(i) if there is at least one overlapping row in SST(j) `

The idea was to DO invoke index.

> 
> Also I think the write amplification could be terrible if you have multiple sstables overlapping each other at their row range.
> 
> SSTable A overlaps with B, which overlaps with C and so on.

If there is such an overlapping then the user was not supposed to use TWCS in the first place and use STCS/ICS instead.


> 
> To be really honest, I think the more appropriate solution is to introduce the concept of runs to TWCS, and that's done by applying ICS to it instead. Better cost benefit (yet more efficient).
> 
> So we fix this issue and #16514 too, killing two birds with one stone.

You can't have it in OSS and, note, this is an OSS GH issue.

**Comment by dani-tweig** [comment]
After the fix of https://github.com/scylladb/scylla-enterprise/issues/3733 is there anything else to be done here?
Can we close this one?

**Comment by vladzcloudius** [comment]
> After the fix of [scylladb/scylla-enterprise#3733](https://github.com/scylladb/scylla-enterprise/issues/3733) is there anything else to be done here? Can we close this one?

@dani-tweig how are these related exactly?
Isn't TWCS still using STCS in each bucket level?

This issue is only going to be resolved when the above had changed.

cc @raphaelsc
